### PR TITLE
download working

### PR DESCRIPTION
**Pull request by golgoth31**
Add ability to download file with progress bar (different from original)

This PR still need some work on tests but opened for discussion

### CODE REVIEW

**Comment by golgoth31** [comment]
closes #27

**Comment by nojima** [comment]
I tried this PR in my local environment. `--download` worked. That's nice :smiley: 

One question, why you use cavaliercoder/grab? It seemed to me that net/http is enough for this requirement.

**Comment by golgoth31** [comment]
Hi, thank: :)
I used grab to be able to take advantage of some advanced features in the future (like resume, checksum or concurrent download)

**Comment by nojima** [comment]
Oh, I see.

Generally speaking, I prefer minimal dependencies. "Minimal" means that all dependencies contribute to the some functionality of httpie-go.

All programs can have bugs and vulnerabilities. The more dependencies we have, the more we need to address bugs and vulnerabilities in dependent libraries. We also need to keep up with breaking changes in these libraries.

Therefore, I want to delay the introduction of the library until it is really needed in the future.

**Comment by golgoth31** [comment]
Ok, I understand the point, I will revert to standard net/http client and only keep a "grab" branch

**Comment by golgoth31** [comment]
Are you ok to abstract fs writing with afero lib ? https://github.com/spf13/afero

**Comment by nojima** [comment]
> I can easily integrate a safe "overwrite" flag in the same time

Sorry. I can't seize the meaning. What is the safe overwrite?

**Comment by golgoth31** [comment]
Writing you the answer make me realize the "safe" overwrite is just writing 2 times the same thing ..... :smile: 
I wrote a simple file writing using os and io :)

**Comment by golgoth31** [comment]
Do you need any other modifications ?

**Comment by nojima** [comment]
I didn't notice that this pull request could be reviewed now. I'll review!

File: output/file.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+	outFile := filepath.Base(fullPath)
+
+	return &FileWriter{
+		FullPath: fullPath,
+		Filename: outFile,
+	}
+}
+
+func checkFile(path string) string {
```
Comment (nojima) [inline output/file.go]: The name `checkFile` did not give me enough information about this function actually doing.
This function's responsibility is giving a new name that does not overlap with existing files.
I think `makeNonOverlappingFilename` might be more informative.

File: output/file.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/vbauerster/mpb/v5/decor"
+)
+
+type FileWriter struct {
+	FullPath string
+	Filename string
+}
+
+func NewFileWriter(url *url.URL, options *Options) *FileWriter {
+	var tempPath string
```
Comment (nojima) [inline output/file.go]: `tempPath`, `tempFile` and `fullPath` can be merged into one variable.

```go
	var fullPath string
	if options.OutputFile == "" {
		fullPath = "./" + filepath.Base(url.Path)
	} else {
		fullPath = options.OutputFile
	}
	if !options.Overwrite {
		fullPath = checkFile(fullPath)
	}
```

File: output/file.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/vbauerster/mpb/v5"
+	"github.com/vbauerster/mpb/v5/decor"
+)
+
+type FileWriter struct {
+	FullPath string
+	Filename string
```
Comment (nojima) [inline output/file.go]: `Filename` seems to be unused.

File: flags/flags.go [hunk 4]
```diff
@@ -110,13 +117,26 @@ func parse(args []string, terminalInfo terminalInfo) ([]string, Usage, *OptionSe
 	if err != nil {
 		return nil, nil, nil, err
 	}
+	if outputOptions.Download {
+		d = time.Duration(0)
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply flags/flags.go]: The download is automatically canceled at the end of the timeout. So I disabled it when downloading.

File: output/file.go (hunk 3 above)
Comment (golgoth31) [reply output/file.go]: Filename is used in main.go, line 116

File: output/plain.go [hunk 5]
```diff
@@ -39,9 +40,17 @@ func (p *PlainPrinter) PrintHeader(header http.Header) error {
 }
 
 func (p *PlainPrinter) PrintBody(body io.Reader, contentType string) error {
+	if body == nil {
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply output/plain.go]: I have got some panic on that but I can't reproduce .... so removed :)

File: flags/flags.go [hunk 6]
```diff
@@ -68,6 +70,11 @@ func parse(args []string, terminalInfo terminalInfo) ([]string, Usage, *OptionSe
 	flagSet.BoolVarLong(&headersFlag, "headers", 'h', "print only the request headers. shortcut for --print=h")
 	flagSet.BoolVarLong(&bodyFlag, "body", 'b', "print only response body. shourtcut for --print=b")
 	flagSet.BoolVarLong(&ignoreStdin, "ignore-stdin", 0, "do not attempt to read stdin")
+	flagSet.BoolVarLong(&outputOptions.Download, "download", 'd', "download file")
+	flagSet.BoolVarLong(&outputOptions.Overwrite, "overwrite", 0, "overwrite existing file")
+	flagSet.BoolVarLong(&http1Flag, "h1", 0, "force HTTP/1.1 protocol")
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply flags/flags.go]: Done.

File: flags/flags.go [hunk 7]
```diff
@@ -110,13 +117,26 @@ func parse(args []string, terminalInfo terminalInfo) ([]string, Usage, *OptionSe
...
+		exchangeOptions.FollowRedirects = true
+	}
 	exchangeOptions.Timeout = d
 
 	// Parse --pretty
 	if err := parsePretty(prettyFlag, terminalInfo.stdoutIsTerminal, &outputOptions); err != nil {
 		return nil, nil, nil, err
 	}
 
+	// Virify SSL
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply flags/flags.go]: Done.

File: main.go [hunk 8]
```diff
@@ -100,20 +100,49 @@ func Exchange(in *input.Input, exchangeOptions *exchange.Options, outputOptions
...
+			if err := printer.PrintHeader(resp.Header); err != nil {
+				return err
+			}
+			writer.Flush()
+		}
+		if outputOptions.PrintResponseBody {
+			if err != nil {
+				return errors.Wrap(err, "print body")
+			}
+			defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply main.go]: Done.

File: main.go [hunk 9]
```diff
@@ -100,20 +100,49 @@ func Exchange(in *input.Input, exchangeOptions *exchange.Options, outputOptions
...
+
+		if err := printer.PrintDownload(resp.ContentLength, file.Filename); err != nil {
 			return err
 		}
 		writer.Flush()
-	}
-	if outputOptions.PrintResponseBody {
-		if err := printer.PrintBody(resp.Body, resp.Header.Get("Content-Type")); err != nil {
+
+		err = file.Download(resp)
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply main.go]: Done.

File: output/plain.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (golgoth31) [reply output/plain.go]: Done.

File: README.md [hunk 10]
```diff
@@ -37,6 +37,18 @@ Request HTTP headers can be specified in the form of `key:value`.
 $ ht -v POST httpbin.org/post X-Foo:foobar
 ```
 
+Disable TLS verification
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply README.md]: Done.

File: output/file.go (hunk 1 above)
Comment (golgoth31) [reply output/file.go]: Done.

File: output/file.go (hunk 2 above)
Comment (golgoth31) [reply output/file.go]: Done.

File: main.go [hunk 11]
```diff
@@ -100,20 +100,49 @@ func Exchange(in *input.Input, exchangeOptions *exchange.Options, outputOptions
...
+			if err := printer.PrintStatusLine(resp.Proto, resp.Status, resp.StatusCode); err != nil {
+				return err
+			}
+			if err := printer.PrintHeader(resp.Header); err != nil {
+				return err
+			}
+			writer.Flush()
+		}
+		if outputOptions.PrintResponseBody {
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply main.go]: Done.

File: flags/flags.go [hunk 12]
```diff
@@ -68,6 +70,11 @@ func parse(args []string, terminalInfo terminalInfo) ([]string, Usage, *OptionSe
 	flagSet.BoolVarLong(&headersFlag, "headers", 'h', "print only the request headers. shortcut for --print=h")
 	flagSet.BoolVarLong(&bodyFlag, "body", 'b', "print only response body. shourtcut for --print=b")
 	flagSet.BoolVarLong(&ignoreStdin, "ignore-stdin", 0, "do not attempt to read stdin")
+	flagSet.BoolVarLong(&outputOptions.Download, "download", 'd', "download file")
+	flagSet.BoolVarLong(&outputOptions.Overwrite, "overwrite", 0, "overwrite existing file")
+	flagSet.BoolVarLong(&http1Flag, "h1", 0, "force HTTP/1.1 protocol")
+	flagSet.StringVarLong(&outputOptions.OutputFile, "output", 'o', "output file")
+	flagSet.StringVarLong(&verifyFlag, "verify", 0, "verify Host SSL certificate ('no' to disable, CA file path otherwise)")
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply flags/flags.go]: Done.

**Comment by golgoth31** [comment]
Thank you for your review ! Sorry for those mistakes, I'm still learning golang :)

File: output/file.go (hunk 3 above)
Comment (nojima) [reply output/file.go]: Oh, I missed it

File: output/file.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+			decor.Elapsed(decor.ET_STYLE_GO, decor.WC{W: 4}),
+			decor.Name(" - "),
+			decor.OnComplete(
+				decor.AverageETA(decor.ET_STYLE_GO, decor.WC{W: 4}), "done",
+			),
+		),
+	)
+
+	// Update progress bar while writing file
+	io.Copy(file, bar.ProxyReader(resp.Body))
```
Comment (nojima) [inline output/file.go]: `io.Copy` returns an error. Please handle it.

Basically, you must always handle errors when functions return errors.
(Functions like `Close()` or `Printf()` are exception.)

File: output/file.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/vbauerster/mpb/v5"
+	"github.com/vbauerster/mpb/v5/decor"
+)
+
+type FileWriter struct {
+	FullPath string
+	Filename string
```
Comment (nojima) [inline output/file.go]: `Filename` is always calculated by `filepath.Base(FullPath)`.
So I think that keeping `Filename` in the struct is redundant.
Please remove `Filename` from struct and create new method `Filename()` to calculate filename.

Comment (nojima) [reply output/file.go]: `FullPath` is not used outside of this module.
So `FullPath` can be private.

File: exchange/client.go [hunk 15]
```diff
@@ -12,10 +15,30 @@ func BuildHTTPClient(options *Options) (*http.Client, error) {
 	if options.FollowRedirects {
 		checkRedirect = nil
 	}
+	tlsConf := &tls.Config{}
+
+	if options.SkipVerify {
+		tlsConf.InsecureSkipVerify = true
+	}
+
+	transp := &http.Transport{
```
Comment (nojima) [inline exchange/client.go]: I found that zero value of `http.Transport` is NOT `DefaultTransport`.
We should explicitly `Clone` the `DefaultTransport` and then overwrite `TLSClientConfig`.

This SO answer would help.
https://stackoverflow.com/a/51848441/5552868

In the `DefaultTrasnport`, `http.ProxyFromEnvironment` is already set. So we can remove the line setting it.

File: exchange/client.go [hunk 16]
```diff
@@ -12,10 +15,30 @@ func BuildHTTPClient(options *Options) (*http.Client, error) {
 	if options.FollowRedirects {
 		checkRedirect = nil
 	}
+	tlsConf := &tls.Config{}
```
Comment (nojima) [inline exchange/client.go]: These lines can be shorten:

```go
tlsConfig := &tls.Config{
    InsecureSkipVerify: options.SkipVerify,
}
```

File: exchange/client.go [hunk 17]
```diff
@@ -12,10 +15,30 @@ func BuildHTTPClient(options *Options) (*http.Client, error) {
...
+	}
+
+	transp := &http.Transport{
+		Proxy:           http.ProxyFromEnvironment,
+		TLSClientConfig: tlsConf,
+		// ForceAttemptHTTP2: true,
+	}
+
+	if !options.ForceHTTP1 {
+		err := http2.ConfigureTransport(transp)
```
Comment (nojima) [inline exchange/client.go]: I think HTTP2 is enabled by default.
So this line does not required.

File: exchange/client.go [hunk 18]
```diff
@@ -12,10 +15,30 @@ func BuildHTTPClient(options *Options) (*http.Client, error) {
...
+	tlsConf := &tls.Config{}
+
+	if options.SkipVerify {
+		tlsConf.InsecureSkipVerify = true
+	}
+
+	transp := &http.Transport{
+		Proxy:           http.ProxyFromEnvironment,
+		TLSClientConfig: tlsConf,
+		// ForceAttemptHTTP2: true,
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply exchange/client.go]: Done.

File: exchange/client.go (hunk 15 above)
Comment (golgoth31) [reply exchange/client.go]: Done.

File: exchange/client.go (hunk 17 above)
Comment (golgoth31) [reply exchange/client.go]: Done. I had to do that with de original http transport, once cloned from the default, it's no more needed.

File: exchange/client.go (hunk 16 above)
Comment (golgoth31) [reply exchange/client.go]: Done.

File: output/file.go [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func NewFileWriter(url *url.URL, options *Options) *FileWriter {
+	var fullPath string
+
+	if options.OutputFile == "" {
+		// tempFile := filepath.Base(url.Path)
```
[the remark this answers is not in the document]
Comment (golgoth31) [reply output/file.go]: Done.

File: output/file.go (hunk 13 above)
Comment (golgoth31) [reply output/file.go]: Done.

File: output/file.go (hunk 14 above)
Comment (golgoth31) [reply output/file.go]: Done.

**Comment by golgoth31** [comment]
Corrected (I hope :crossed_fingers: )

File: flags/flags.go [hunk 20]
```diff
@@ -110,13 +117,32 @@ func parse(args []string, terminalInfo terminalInfo) ([]string, Usage, *OptionSe
...
+		exchangeOptions.SkipVerify = true
+	case "yes":
+	case "":
+		exchangeOptions.SkipVerify = false
+	default:
+		return nil, nil, nil, fmt.Errorf("%s", "Verify flag must be 'yes' or 'no'")
+	}
+
+	// Force HTTP/1.1
+	exchangeOptions.ForceHTTP1 = http1Flag
```
Comment (nojima) [inline flags/flags.go]: ForceHTTP1 flag is no longer used. Please remove this flag.

File: flags/flags.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nojima) [reply flags/flags.go]: Oh, I see.

File: flags/flags.go (hunk 20 above)
Comment (golgoth31) [reply flags/flags.go]: You're right but I think being able to test h2 and http1 could be really useful (it was in my case). The original httpie can't do that natively but curl yes by adding "h2" flag. I'm looking for a solution. :)

Comment (golgoth31) [reply flags/flags.go]: Working !

❯ go run ./cmd/ht -v -F --http1 https://httpbin.org
GET / HTTP/1.1
Accept-Encoding: gzip
Host: 
User-Agent: httpie-go/0.6.0


HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Connection: keep-alive
Content-Length: 9593
Content-Type: text/html; charset=utf-8
Date: Fri, 20 Mar 2020 09:11:07 GMT
Server: gunicorn/19.9.0

❯ go run ./cmd/ht -v -F https://httpbin.org 
GET / HTTP/1.1
Accept-Encoding: gzip
Host: 
User-Agent: httpie-go/0.6.0


HTTP/2.0 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Content-Length: 9593
Content-Type: text/html; charset=utf-8
Date: Fri, 20 Mar 2020 09:11:22 GMT
Server: gunicorn/19.9.0

**Comment by golgoth31** [comment]
❯ go test ./...                                                                                                           
?   	github.com/nojima/httpie-go	[no test files]
?   	github.com/nojima/httpie-go/cmd/ht	[no test files]
ok  	github.com/nojima/httpie-go/exchange	(cached)
ok  	github.com/nojima/httpie-go/flags	(cached)
ok  	github.com/nojima/httpie-go/input	(cached)
ok  	github.com/nojima/httpie-go/output	(cached)
?   	github.com/nojima/httpie-go/version	[no test files]

File: exchange/client.go [hunk 21]
```diff
@@ -13,9 +14,18 @@ func BuildHTTPClient(options *Options) (*http.Client, error) {
 		checkRedirect = nil
 	}
 
+	transp := http.DefaultTransport.(*http.Transport).Clone()
+	transp.TLSClientConfig.InsecureSkipVerify = options.SkipVerify
+	if options.ForceHTTP1 {
+		transp.TLSClientConfig.NextProtos = []string{"http1"}
```
Comment (nojima) [inline exchange/client.go]: `http1` does not seem to be correct.
I looked at Go source code, but  could not find `http1` used as a value of `NextProtos`.
Perhaps `http/1.1` is right.

I investigated the semantics of `NextProtos` and `TLSNextProto`, and how these configurations affects others.
But the configurations around HTTP and TLS were too complicated to understand.
I'm nervous whether this change is breaking anything I don't expect.

So I want to simply drop `--http1` flag, and allow Go runtime to choose protocol.

File: exchange/options.go [hunk 22]
```diff
@@ -6,6 +6,9 @@ type Options struct {
 	Timeout         time.Duration
 	FollowRedirects bool
 	Auth            AuthOptions
+	SkipVerify      bool
+	ForceHTTP1      bool
+	Dst             string
```
Comment (nojima) [inline exchange/options.go]: Unused?

Comment (golgoth31) [reply exchange/options.go]: Fixed.

File: exchange/client.go (hunk 21 above)
Comment (golgoth31) [reply exchange/client.go]: As previously mentioned, I used this switch to test envoy proxy and this was really useful. The go documentation is not really clear but: "Programs that must disable HTTP/2 can do so by setting Transport.TLSNextProto (for clients)" was not enough so I had to force the Nextproto something different then "h2".

**Comment by nojima** [comment]
@golgoth31 
I merged your PR. Thank you for your hard work :+1:

**Comment by golgoth31** [comment]
Thank you for your time !

**Pull request merged**